Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -0,0 +1,311 @@
+
+            context.Publish(
+                new StandardOutputValueProduced(
+                    output,
+                    command,
+                    formattedValues));
+        }
+
+        private void PublishStreamRecord(
+            object output,
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            context.Publish(
+                new DisplayedValueProduced(
+                    output,
+                    command,
+                    FormattedValue.FromObject(output)));
+        }
+
+        private void PublishError(
+            string error,
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            var formattedValues = new List<FormattedValue>
+            {
+                new FormattedValue(
+                    PlainTextFormatter.MimeType, error)
+            };
+            
+            context.Publish(
+                new StandardErrorValueProduced(
+                    error,
+                    command,
+                    formattedValues));
+        }
+
+        private async Task<IEnumerable<CompletionItem>> GetCompletionList(
+            string code,
+            int cursorPosition)
+        {
+            await _runspaceSemaphore.WaitAsync();
+
+            try
+            {
+                using (var ps = System.Management.Automation.PowerShell.Create(_runspace))
+                {
+                    CommandCompletion completion = CommandCompletion.CompleteInput(code, cursorPosition, null, ps);
+
+                    return completion.CompletionMatches.Select(c => new CompletionItem(
+                        displayText: c.CompletionText,
+                        kind: c.ResultType.ToString(),
+                        documentation: c.ToolTip
+                    ));
+                }
+            }
+            finally
+            {
+                _runspaceSemaphore.Release();
+            }
+        }
+
+        public async Task LoadExtensionsFromDirectory(
+            DirectoryInfo directory,
+            KernelInvocationContext context)
+        {
+            var extensionsDirectory =
+                new DirectoryInfo(
+                    Path.Combine(
+                        directory.FullName,
+                        "interactive-extensions",
+                        "dotnet",
+                        "cs"));
+
+            await new KernelExtensionAssemblyLoader().LoadFromAssembliesInDirectory(
+                extensionsDirectory,
+                context.HandlingKernel,
+                context);
+        }
+
+        private void RegisterPowerShellStreams(
+            System.Management.Automation.PowerShell powerShell,
+            KernelInvocationContext context,
+            IKernelCommand command)
+        {
+            var streamHandler = new StreamHandler(context, command);
+            powerShell.Streams.Debug.DataAdding += streamHandler.DebugDataAdding;
```

Comment:
Are you unhooking these events anywhere?

Please make sure to dispose anything that needs disposing.

Reply:
These events are hooked up to the PowerShell object that gets disposed via the `using(...)`. I could unhook them if you'd like but it shouldn't be needed.

Reply:
It should usually be fine. We just spent too much time tracking down a memory leak resulting from a dangling object reference, which could conceivably happen here.

Reply:
I can also clean them up 👍